### PR TITLE
dev-python/argon2_cffi: New ebuild

### PR DESCRIPTION
**Pull request by seifertm**
Adds an ebuild for Python bindings of the argon2 password hashing algorithms. Please see the linked bug for more information.

I intend to maintain this package, because I am using it productively. Since this would be my first ebuild as a proxied maintainer, there is no maintainer bug in the issue tracker for me.

Closes: https://bugs.gentoo.org/643904
Signed-off-by: Michael Seifert <m.seifert@digitalernachschub.de>

### CODE REVIEW

**Comment by seifertm** [comment]
Is @gentoo/python interested in co-maintaining this package? I feel like argon2_cffi is valuable addition for the Python ecosystem in Gentoo. For example the PyPI mirror devpi uses argon2_cffi (via passlib).

**Comment by mgorny** [comment]
Sure, you can add us as backup maintainer. You should stay as the primary maintainer, though.

**Comment by seifertm** [comment]
Thanks! I added the Python project as second maintainer in `metadata.xml`.

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+CDEPEND="
+	~app-crypt/argon2-20171227:=
+	dev-python/six[${PYTHON_USEDEP}]
+	$(python_gen_cond_dep '>=dev-python/cffi-1.0.0[${PYTHON_USEDEP}]' 'python*')
+	$(python_gen_cond_dep 'dev-python/enum34[${PYTHON_USEDEP}]' 'python2_7' 'pypy')
+"
+RDEPEND="${CDEPEND}"
+DEPEND="${CDEPEND}
+	dev-python/setuptools[${PYTHON_USEDEP}]
+	doc? (	dev-python/sphinx[${PYTHON_USEDEP}] )
```
Comment (mgorny) [inline dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: Please don't use tabs in middle of the line.

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+LICENSE="MIT"
+
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="doc test"
+
+CDEPEND="
+	~app-crypt/argon2-20171227:=
+	dev-python/six[${PYTHON_USEDEP}]
+	$(python_gen_cond_dep '>=dev-python/cffi-1.0.0[${PYTHON_USEDEP}]' 'python*')
```
Comment (mgorny) [inline dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: Use `virtual/python-cffi` instead.

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	$(python_gen_cond_dep 'dev-python/enum34[${PYTHON_USEDEP}]' 'python2_7' 'pypy')
```
Comment (mgorny) [inline dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: `virtual/python-enum34`? ;-)

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+PYTHON_COMPAT=( python{2_7,3_{4,5,6,7}} pypy pypy3 )
+inherit distutils-r1
+
+DESCRIPTION="CFFI bindings to the Argon2 password hashing library"
+HOMEPAGE="https://github.com/hynek/argon2_cffi"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+
+LICENSE="MIT"
+
```
Comment (mgorny) [inline dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: Why this empty line?

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="doc test"
+
+CDEPEND="
```
Comment (mgorny) [inline dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: Please don't name it CDEPEND. We might be crazy enough to use CDEPEND for something else in a future EAPi, and it's easy to miss the accidental collision.

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+		dev-python/coverage[${PYTHON_USEDEP}]
+		dev-python/hypothesis[${PYTHON_USEDEP}]
+		>=dev-python/pytest-3.0[${PYTHON_USEDEP}]
+	)
+"
+
+DOCS=( AUTHORS.rst CHANGELOG.rst FAQ.rst README.rst )
+
+python_compile() {
+	export ARGON2_CFFI_USE_SYSTEM=1
```
Comment (mgorny) [inline dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: You can use `python_configure_all()` phase for this instead. It doesn't do anything by default, so you'd avoid having to call distutils-r1_something.

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	distutils-r1_python_compile
+}
+
+python_compile_all() {
+	use doc && emake -C docs html
+}
+
+python_test() {
+	py.test || die "Tests failed with ${EPYTHON}"
```
Comment (mgorny) [inline dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: `pytest -vv` is the modern syntax, please.

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild (hunk 3 above)
Comment (seifertm) [reply dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: Didn't know these virtuals exist. Thanks for the hint!

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild (hunk 5 above)
Comment (seifertm) [reply dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: Fair point. I used to write `RDEPEND="<python deps>"` and `DEPEND="${RDEPEND} <build deps>"`. Other devs disliked that version as well. Is there any good practice you can recommend for Python ebuilds?

Comment (mgorny) [reply dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: I have no clue why anybody would oppose using RDEPEND+DEPEND as you said. Generally if one is subset of the other, declare the smaller one first, then use it in the other. If dep groups aren't exact subset, either use `COMMON_DEPEND` or something like `DEPEND="..."; RDEPEND="${DEPEND} ..."; DEPEND+="..."`.

File: dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	dev-python/six[${PYTHON_USEDEP}]
+	virtual/python-cffi[${PYTHON_USEDEP}]
+	virtual/python-enum34[${PYTHON_USEDEP}]
+"
+RDEPEND="${COMMON_DEPEND}"
+DEPEND="${COMMON_DEPEND}
+	dev-python/setuptools[${PYTHON_USEDEP}]
+	doc? ( dev-python/sphinx[${PYTHON_USEDEP}] )
+	test? (
+		dev-python/coverage[${PYTHON_USEDEP}]
```
Comment (mgorny) [inline dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: So I've just tested it and it passes just fine without coverage.

Comment (seifertm) [reply dev-python/argon2_cffi/argon2_cffi-18.3.0.ebuild]: Same here. Coverage is listed in `setup.py`, but it is only used in `tox.ini` which we don't need.
I removed it from the test dependencies.

**Pull request closed without merging**